compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -527,6 +534,160 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// Combines all the reasons for 2229 migrations
+    fn compute_2229_migrations_reasons(
+        &self,
+        set_reasons: FxHashSet<&str>
+    ) -> String {
+        let mut reasons = String::new();
+        if set_reasons.contains("Clone") {
+            reasons = format!("'Clone'");
+        }
+
+        if set_reasons.contains("Sync") {
+            if !reasons.is_empty() {
+                reasons = format!("{}, ", reasons);
+            }
+            reasons = format!("{}'Sync'", reasons);
+        }
+
+        if set_reasons.contains("Unpin") {
+            if !reasons.is_empty() {
+                reasons = format!("{}, ", reasons);
+            }
+            reasons = format!("{}'Unpin'", reasons);
+        }
+
+        if set_reasons.contains("Clone")
+            || set_reasons.contains("Sync")
+            || set_reasons.contains("Unpin")
+        {
+            reasons = format!("{} trait implementation", reasons);
+        }
+
+        if set_reasons.contains("drop order") {
+            if !reasons.is_empty() {
+                reasons = format!("{}, and ", reasons);
+            }
+            reasons = format!("{}drop order", reasons);
+        }
+
+        reasons
+    }
+
+    /// Figures out the list of root variables (and their types) that aren't completely
+    /// captured by the closure when `capture_disjoint_fields` is enabled and auto-traits
+    /// differ between the root variable and the captured paths.
+    ///
+    /// The output list would include a root variable if:
+    /// - It would have been captured into the closure when `capture_disjoint_fields` wasn't
+    ///   enabled, **and**
+    /// - It wasn't completely captured by the closure, **and**
+    /// - One of the paths captured does not implement all the auto-traits its root variable
+    ///   implements.
+    fn compute_2229_migrations_for_trait(
+        &self,
+        min_captures: Option<&ty::RootVariableMinCaptureList<'tcx>>,
+        closure_clause: hir::CaptureBy,
+        var_hir_id: hir::HirId,
+    ) -> Option<(hir::HirId, FxHashSet<&str>)> {
+        use crate::rustc_middle::ty::ToPredicate;
+        use crate::rustc_middle::ty::WithConstness;
+        use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+
+        let ty = self.infcx.resolve_vars_if_possible(self.node_ty(var_hir_id));
+
+        let tcx = self.infcx.tcx;
+
+        let cause = ObligationCause::misc(self.tcx.hir().span(var_hir_id), self.body_id);
+
+        // <ty as Clone>
+        let clone_trait_ref = TraitRef {
```

Comment:
maybe create a vector of tuple (trait_def_id, reason_string) and then iterator over them?

Reply:
Don't think it would help a lot, maybe creating a function instead? I may come back to this once I am in a more final version